x-pack/plugins/maps/public/connected_components/right_side_controls/layer_control/layer_toc/layer_toc.tsx

```diff
@@ -29,60 +52,201 @@ export class LayerTOC extends Component<Props> {
+      ? this._getDepth(parent, depth + 1)
+      : { depth, showInTOC: false };
+  }
+
+  _getDroppableClass() {
+    if (!this.state.sourceLayer) {
+      // nothing is dragged
+      return '';
+    }
+
+    if (this.state.isOwnAncestor) {
+      return 'mapLayerToc-droppable-dropNotAllowed';
+    }
+
+    if (this.state.combineLayer) {
+      return 'mapLayerToc-droppable-isCombining';
+    }
+
+    return 'mapLayerToc-droppable-isDragging';
+  }
+
   _renderLayers() {
-    // Reverse layer list so first layer drawn on map is at the bottom and
-    // last layer drawn on map is at the top.
-    const reverseLayerList = [...this.props.layerList].reverse();
+    const tocEntryList = this.props.layerList
+      .map((layer, index) => {
+        return {
+          ...this._getDepth(layer, 0),
+          draggableIndex: this._reverseIndex(index),
+          layer,
+        };
+      })
+      .filter(({ showInTOC }) => {
+        return showInTOC;
+      })
+      // Reverse layer list so first layer drawn on map is at the bottom and
+      // last layer drawn on map is at the top.
+      .reverse();
 
     if (this.props.isReadOnly) {
-      return reverseLayerList.map((layer) => {
-        return <TOCEntry key={layer.getId()} layer={layer} />;
+      return tocEntryList.map(({ depth, layer }) => {
+        return <TOCEntry key={layer.getId()} depth={depth} layer={layer} />;
       });
     }
 
     return (
-      <EuiDragDropContext onDragEnd={this._onDragEnd}>
-        <EuiDroppable droppableId="mapLayerTOC" spacing="none">
-          {(droppableProvided, snapshot) => {
-            const tocEntries = reverseLayerList.map((layer, idx: number) => (
+      <EuiDragDropContext
+        onDragStart={this._onDragStart}
+        onDragUpdate={this._onDragUpdate}
+        onDragEnd={this._onDragEnd}
+      >
+        <EuiDroppable
+          droppableId="mapLayerTOC"
+          spacing="none"
+          isCombineEnabled={!this.state.isOwnAncestor}
+          className={this._getDroppableClass()}
+        >
+          {(droppableProvided, droppableSnapshot) => {
+            const tocEntries = tocEntryList.map(({ draggableIndex, depth, layer }) => (
               <EuiDraggable
                 spacing="none"
                 key={layer.getId()}
-                index={idx}
+                index={draggableIndex}
                 draggableId={layer.getId()}
                 customDragHandle={true}
                 disableInteractiveElementBlocking // Allows button to be drag handle
               >
-                {(provided, state) => (
-                  <TOCEntry
-                    layer={layer}
-                    dragHandleProps={provided.dragHandleProps}
-                    isDragging={state.isDragging}
-                    isDraggingOver={snapshot.isDraggingOver}
-                  />
-                )}
+                {(draggableProvided, draggableSnapshot) => {
+                  return (
+                    <TOCEntry
```

Comment:
nit. It would be nice if the child layers in the layer group were included in the draggable content. It seems weird to see the "Layer group" component move separately than its children.


https://user-images.githubusercontent.com/1638483/195178095-59c23c9c-d26f-4214-ab9f-0eb768caf203.mp4

Reply:
> It would be nice if the child layers in the layer group were included in the draggable content

This was not technically possible. I tried to allow dragging a layer group with children by nesting `EuiDraggable` components where a layer group would be a `EuiDraggable` component containing all children layers and the children would be there own `EuiDraggable` components within layer group.  React dnd did not like nested `EuiDraggable` and did not work.

Reply:
I wonder then if we could collapse the layer details when dragging a layer group? Then when dragging ends, re-open the layer details. Again, not a big deal, but might be nicer user experience.

Reply:
> I wonder then if we could collapse the layer details when dragging a layer group? Then when dragging ends, re-open the layer details. Again, not a big deal, but might be nicer user experience.

That is a good idea but unfortunately, react dnd has buggy behavior when the container size changes while dragging.